### PR TITLE
Introduce a cluster option for workshops

### PR DESCRIPTION
**Pull request by termlen0**
##### SUMMARY

This PR introduces the option of running any of the workshops with Tower as a cluster.
- Users will need to add the `create_cluster: yes` option to their vars file for this to work.
##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- provisioner

##### ADDITIONAL INFORMATION

This option will allow SAs to build even better demos around scaling/loadbalancing(RBAC). 
Additionally, when provisioned for workshops, it will help highlight Tower scaling/clustering features to the customer.

CC: @IPvSean 

```paste below

```

### CODE REVIEW

File: provisioner/roles/manage_ec2_instances/tasks/provision.yml [hunk 1]
```diff
@@ -28,51 +28,12 @@
...
-      Index: "{{ item[0] }}"
-      Student: "student{{item.0 + 1}}"
-      dns: "{% if dns_type != 'none' %}student{{ item.0 + 1 }}.{{ ec2_name_prefix|lower }}.{{ workshop_dns_zone }} {% else %} none {% endif %}"
-      launch_time: "{{ item.1.launch_time }}"
-  with_indexed_items:
-    - "{{ control_output.instances }}"
-  when: control_output.instance_ids is not none
-
+- name: Create the control clusters
+  include_tasks: cluster_instances.yaml
```
Comment (cigamit) [inline provisioner/roles/manage_ec2_instances/tasks/provision.yml:32]: Can you change this file extension? Everything else in the project uses .yml

File: provisioner/group_vars/all/vpc_rules.yml [hunk 2]
```diff
@@ -60,6 +60,26 @@ workshops:
...
+      - proto: tcp
+        to_port: 5432
+        from_port: 5432
+        cidr_ip: 0.0.0.0/0
+        rule_desc: Cluster option DB port
+      - proto: tcp
+        to_port: 4369
+        from_port: 4369
+        cidr_ip: 0.0.0.0/0
+        rule_desc: Cluster option RabbitMQ
```
Comment (cigamit) [inline provisioner/group_vars/all/vpc_rules.yml:72]: 3.7 no longer uses RabbitMQ, but even so I wouldn't open it to the world.  If I recall for early versions of the workshop that used completely wide open firewall rules, RabbitMQ was causing the boxes to get exploited.  Better safe than sorry.

Instead I would modify the rule to allow only internal IPs, and modify the tower setup inventory to reference the internal IPs also.

**Comment by cigamit** [comment]
I made a few comments inline just from things I saw.  I like this option, but a few things I would fix.  I will run a few tests tonight to see what else I notice.  Sean isn't a fan of loops for building VMs, so we may have to think outside the box on how to do them all at once, but tag each one separately for which node it is.

File: provisioner/group_vars/all/vpc_rules.yml [hunk 3]
```diff
@@ -60,6 +60,26 @@ workshops:
         from_port: 514
         cidr_ip: 0.0.0.0/0
         rule_desc: WinRM
+      - proto: tcp
+        to_port: 5432
```
Comment (cigamit) [inline provisioner/group_vars/all/vpc_rules.yml:64]: Same here, I would make Tower use the internal IP to access the DB versus the Public IP.  That way we don't have to open Postgres to the world.

File: provisioner/roles/manage_ec2_instances/tasks/provision.yml (hunk 1 above)
Comment (termlen0) [reply provisioner/roles/manage_ec2_instances/tasks/provision.yml]: Updated

File: provisioner/group_vars/all/vpc_rules.yml (hunk 2 above)
Comment (termlen0) [reply provisioner/group_vars/all/vpc_rules.yml:72]: Updated to use internal IPs

File: provisioner/group_vars/all/vpc_rules.yml (hunk 3 above)
Comment (termlen0) [reply provisioner/group_vars/all/vpc_rules.yml:64]: Updated to use internal IPs

**Comment by cloin** [comment]
Fails in security workshop deploy and rhel-verify and I think that's fine:
```
[2020-05-29T13:36:37.658Z] TASK [manage_ec2_instances : provision workshop instances] *********************

[2020-05-29T13:36:37.658Z] ERROR! Unexpected Exception, this is probably a bug: 'NoneType' object has no attribute 'rfind'

[2020-05-29T13:36:37.658Z] to see the full traceback, use -vvv
```

```
[2020-05-29T14:06:09.263Z] TASK [Test access by exporting assets] *****************************************

[2020-05-29T14:06:09.263Z] skipping: [student2-node1]

[2020-05-29T14:06:09.263Z] skipping: [student2-node2]

[2020-05-29T14:06:09.263Z] skipping: [student2-node3]

[2020-05-29T14:06:09.531Z] skipping: [student1-node1]

[2020-05-29T14:06:09.531Z] skipping: [student1-node3]

[2020-05-29T14:06:09.531Z] skipping: [student1-node2]

[2020-05-29T14:06:10.468Z] fatal: [student2-ansible-1]: FAILED! => changed=false 

[2020-05-29T14:06:10.468Z]   assets: null

[2020-05-29T14:06:10.468Z]   message: |-

[2020-05-29T14:06:10.468Z]     There was a network error of some kind trying to connect to Tower.

[2020-05-29T14:06:10.468Z]   

[2020-05-29T14:06:10.468Z]     The most common  reason for this is a settings issue; is your "host" value in `tower-cli config` correct?

[2020-05-29T14:06:10.468Z]     Right now it is: "student2-1.tqe-rhel-tower370-PR-891-3.rhdemo.io".

[2020-05-29T14:06:10.468Z]   msg: Receive Failed

[2020-05-29T14:06:10.468Z] fatal: [student1-ansible-1]: FAILED! => changed=false 

[2020-05-29T14:06:10.468Z]   assets: null

[2020-05-29T14:06:10.468Z]   message: |-

[2020-05-29T14:06:10.468Z]     There was a network error of some kind trying to connect to Tower.

[2020-05-29T14:06:10.468Z]   

[2020-05-29T14:06:10.468Z]     The most common  reason for this is a settings issue; is your "host" value in `tower-cli config` correct?

[2020-05-29T14:06:10.468Z]     Right now it is: "student1-1.tqe-rhel-tower370-PR-891-3.rhdemo.io".

[2020-05-29T14:06:10.468Z]   msg: Receive Failed
```

**Comment by cloin** [comment]
Recheck

**Comment by termlen0** [comment]
@cloin / @Spredzy Let me know what I can do to help move this PR forward.

File: provisioner/group_vars/all/vpc_rules.yml (hunk 2 above)
Comment (liquidat) [reply provisioner/group_vars/all/vpc_rules.yml:72]: Do we still need the rules if we only use internal IPs?

File: provisioner/group_vars/all/vpc_rules.yml [hunk 4]
```diff
@@ -60,6 +60,26 @@ workshops:
...
+      - proto: tcp
+        to_port: 25672
+        from_port: 25672
+        cidr_ip: 172.16.0.0/14
+        rule_desc: Cluster option RabbitMQ
+      - proto: tcp
+        to_port: 5672
+        from_port: 5672
+        cidr_ip: 172.16.0.0/14
+        rule_desc: Cluster option RabbitMQ
```
Comment (liquidat) [inline provisioner/group_vars/all/vpc_rules.yml:82]: As mentioned above, we don't have RabbitMQ anymore, so please remove the RabbitMQ rules.

**Comment by cloin** [comment]
The check failures don’t seem to be related to changes on this PR. @liquidat how do you feel about this PR?

File: provisioner/roles/control_node/templates/tower_cluster_install.j2 [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+pg_database='awx'
+pg_username='awx'
+pg_password='{{admin_password}}'
+
+rabbitmq_port=5672
+rabbitmq_vhost=tower
+rabbitmq_username=tower
+rabbitmq_password='{{admin_password}}'
+rabbitmq_cookie=cookiemonster
+
```
Comment (liquidat) [inline provisioner/roles/control_node/templates/tower_cluster_install.j2:30]: We don't use RabbitMQ with Tower 3.7 anymore, can you please update this to Tower 3.7?

File: provisioner/roles/manage_ec2_instances/templates/etchosts/etchosts_rhel.j2 [hunk 6]
```diff
@@ -21,6 +21,6 @@
 
 {% for vm in ansible_node_facts.instances %}
 {% if 'student' + item == vm.tags.Student %}
-{{ vm.public_ip_address }} {{vm.tags.Student}}.{{ec2_name_prefix|lower}}.{{workshop_dns_zone}} {{ vm.tags.short_name }}
+{{ vm.private_ip_address }} {{vm.tags.Student}}.{{ec2_name_prefix|lower}}.{{workshop_dns_zone}} {{ vm.tags.short_name }}
```
Comment (liquidat) [inline provisioner/roles/manage_ec2_instances/templates/etchosts/etchosts_rhel.j2:24]: So, with this the nodes don't have public IPs anymore, am I right? We need confirmation for each workshop individually if the exercises really work with private IPs.

**Comment by liquidat** [comment]
@goetzrieger What do you think? Maybe we can start the advanced Tower lab right off of this RPR? Could simplify a lot of work!

**Comment by termlen0** [comment]
I’m counting on our tests. I verified manually, the linux and network workshops.

**Comment by termlen0** [comment]
As for Tower 3.7, these rules will allow workshops to be spun up with older versions.  As an SA sometimes we might use this provisioner to simulate customer environments. This does not affect deployment of 3.7

File: provisioner/roles/manage_ec2_instances/templates/etchosts/etchosts_rhel.j2 (hunk 6 above)
Comment (goetzrieger) [reply provisioner/roles/manage_ec2_instances/templates/etchosts/etchosts_rhel.j2:24]: And we should get away from using the IPs in favor of using the (/etc/hosts) names anyway... ;)

**Comment by goetzrieger** [comment]
First for @termlen0 : This is something we *really* need. @cbolz and me built something like this for Summit to run our Advanced Tower lab on (*advanced_tower* branch). But we didn't find the time to get it into devel (not talking about master) proper so it would be a lot of work now (there where a lot of changes in the mean time...). But we need a clustered env in master to get the lab into RHPDS...

I just had the time for a brief look into what you have done but will try to test, maybe @cbolz can have a look, too.

It would be great to get this working and into devel/master. The Advanced Tower lab itself doesn't have a lot of requirements regarding the lab env but a three-node cluster, but we need to check.

**Comment by liquidat** [comment]
> As for Tower 3.7, these rules will allow workshops to be spun up with older versions. As an SA sometimes we might use this provisioner to simulate customer environments.

Maybe, but it would be news to me that this is a desired feature of the workshops?!

And while it does not effect the actual deployment, it introduces an entire set of legacy code: rabbitmq configuration in the inventory, various firewall rules, etc.

That is not a deal breaker for me, but I would very much prefer to focus on up2date Tower releases. We also do not cater to people anymore who want to use RHEL 7, or other older Tower or Ansible releases.

**Comment by termlen0** [comment]
Thanks for this. I’ll refactor next week to include #892 and update the landing page template.

**Comment by goetzrieger** [comment]
Some more thoughts:

- To avoid creating instances in a loop the most pragmatic way might be to just include a cluster_instances.yml or a single_instance.yml (or so) file in manage_ec2_instances
- As far as I can see you still have all control nodes in the control_nodes group and then you separate tasks by running them either on ansible-1 or ansible-2 to ansible-4 (instead of using groups) in provision_lab.yml. IMHO this is asking for trouble, the group control_nodes could accidentally be used somewhere else, renaming nodes is hard etc). Having a dedicated group for the cluster nodes might be cleaner. Disclaimer: we did this for the Summit env and it gave us loads of fun/pain.   
- This is more cosmetic, but I would like names like tower-1, tower-2 better so students don't get confused (too easily ;).

I know it's hard to get clustering into the Workshops as unintrusively as possible. I'm happy to help if I can.

**Comment by termlen0** [comment]
Per review from @goetzrieger , I've re-based to accommodate PR  #892. I've also updated the landing page J2 to only display the main control node details. Tested with RHEL workshop.

**Comment by termlen0** [comment]
> Some more thoughts:
> 
>     * To avoid creating instances in a loop the most pragmatic way might be to just include a cluster_instances.yml or a single_instance.yml (or so) file in manage_ec2_instances
This is what is currently being done. See: 
```
- name: Create the control clusters
  include_tasks: cluster_instances.yml
  loop: "{{ range(1, control_nodes|default(1) + 1 ) | list }}"
  loop_control:
    loop_var: sequence
```
> 
>    IMHO this is asking for trouble, the group control_nodes could accidentally be used somewhere else, renaming nodes is hard etc. Having a dedicated group for the cluster nodes might be cleaner. Disclaimer: we did this for the Summit env and it gave us loads of fun/pain.
> 
While I agree that it might be a cleaner approach, I fail to see how using the control group for anything else can cause an issue. In the spirit of getting a cluster option in place, I suggest we table this for a future PR.
>     * This is more cosmetic, but I would like names like tower-1, tower-2 better so students don't get confused (too easily ;).
> 
I'm personally not opinionated one way or the other about the naming. Will let others comment on it. But again, not a show-stopper for this PR, IMO.

**Comment by cbolz** [comment]
I like the approach - it's less intrusive then what we hacked together for Summit.

Personally, I would probably split out changes like switching from private to public IPs and upgrading to 3.7 into separate PR's, but I guess that's something up for debate and is just different ways of working.

I also agree with @liquidat that the purpose of the workshop is to have the latest ansible releases and not to carry technical debt to support all sorts of old releases. IMHO that's out of scope for this project.

File: provisioner/group_vars/all/vpc_rules.yml (hunk 2 above)
Comment (liquidat) [reply provisioner/group_vars/all/vpc_rules.yml:72]: @termlen0 See above, do we really need these rules publicly accessible to everyone?

**Comment by termlen0** [comment]
Roland,
You might be replying to an older comment. The PR was updated to use private internal IPs (for the scope of this pr)

**Comment by goetzrieger** [comment]
Just tested with the recent version of your repo/branch, looks good apart from the /etc/hosts file:

```
54.196.216.234 node1
18.212.217.4 node2
100.24.68.189 node3

172.16.93.12 student1.gricluster1.rhdemo.io ansible-1
172.16.158.60 student1.gricluster1.rhdemo.io ansible-3
172.16.77.101 student1.gricluster1.rhdemo.io ansible-4
172.16.18.194 student1.gricluster1.rhdemo.io ansible-2
```

So internal/external names/IPs are kind of mixed up. I didn't check if this is the same way with the non-cluster lab, I'd guess so. But at least it's only one host then. ;)

I think we should take care of this and look at the other labs, too. I'm thinking about rewriting the Linux Automation lab guide to use the /etc/hosts names instead of fiddling with IPs, something that is always confusing students. I did this for the Tower lab at Summit and it worked fine. But /etc/hosts has to be trustworthy then.

**Comment by termlen0** [comment]
Network workshop requires public IP for end points. The control nodes can still be private IPs. The name resolution in your output looks good to me.

**Comment by goetzrieger** [comment]
I agree it's working this way, but in IMO it doesn't make a lot of sense to have the same external hostname pointing to 4 different internal IPs. Even if it's cosmetic there might be trouble later. And the managed nodes in the RHEL workshop could happily use internal IPs, but I admit it's not really important.

On a different note in the Tower Advanced lab we use two "local" managed nodes and two more for the isolated node exercise (we didn't bother with a separate network segment, so not really isolated). We called the nodes isonode and remote. But I can take care of this later.

So @cloin @liquidat what is really keeping this from being merged into devel?

**Comment by goetzrieger** [comment]
> > IMHO this is asking for trouble, the group control_nodes could accidentally be used somewhere else, renaming nodes is hard etc. Having a dedicated group for the cluster nodes might be cleaner. Disclaimer: we did this for the Summit env and it gave us loads of fun/pain.
> 
> While I agree that it might be a cleaner approach, I fail to see how using the control group for anything else can cause an issue. In the spirit of getting a cluster option in place, I suggest we table this for a future PR.

If anybody is doing something (installing software, whatever) on the control node group without limiting to ansible-1 (like introduced with the cluster feature) this will run on all cluster nodes the way it's implemented now. And introducing a dedicated group later is quite a lot of work/changes, we've been there. 

I might be overly cautious, though. What do the others think about this? I'd like to see the clustering integrated rather sooner than later, too.

File: provisioner/group_vars/all/vpc_rules.yml (hunk 2 above)
Comment (termlen0) [reply provisioner/group_vars/all/vpc_rules.yml]: Agree. Removed these rules.

**Comment by goetzrieger** [comment]
@cloin @liquidat ?

**Comment by liquidat** [comment]
@termlen0 This is shaping up really nicely, I love it! One small thing missing: we need a sample vars file or at least entry for each new option we bring in, can you add this?

Also, while I would say @goetzrieger is a bit too cautious with the thoughts around control_nodes, we should at leas make sure that the other labs are working on it. So we need to fix these lines where stuff is install on control_nodes:

- https://github.com/ansible/workshops/blob/2ff691f0471a0f3f2d9a4753dc395510b9e98471/provisioner/windows.yml#L17
- https://github.com/ansible/workshops/blob/2ff691f0471a0f3f2d9a4753dc395510b9e98471/provisioner/roles/workshop_attendance/templates/workshop.sql.j2#L19
- https://github.com/ansible/workshops/blob/2ff691f0471a0f3f2d9a4753dc395510b9e98471/provisioner/security.yml#L111

Number one and three can just be rewritten to the ansible-1, but I am not sure about number 2, maybe @cloin can help?

**Comment by goetzrieger** [comment]
Ajay, I know this is a bit selfish, but while you are at it:

Could we have four managed nodes, ideally _node1_/_node2_, then _isonode_ and _remotenode_?

Then this environment would line up with the Advanced Tower lab [1] perfectly. If this is asked too much I'll give it a shot later.


[1] https://people.redhat.com/grieger/summit2020_labs/ansible-tower-advanced/8-isolated-nodes/

**Comment by termlen0** [comment]
> Ajay, I know this is a bit selfish, but while you are at it:
> 
> Could we have four managed nodes, ideally _node1_/_node2_, then _isonode_ and _remotenode_?
> 
> Then this environment would line up with the Advanced Tower lab [1] perfectly. If this is asked too much I'll give it a shot later.
> 
> [1] https://people.redhat.com/grieger/summit2020_labs/ansible-tower-advanced/8-isolated-nodes/
I'll try. For now, path of least resistance for me to get this PR into devel is to try and address @liquidat 's 3 items. I'm going to try and get that in first and once the PR is merged into devel, start a new feature branch to refactor it per your suggestion. Hope that works. :)
Cheers.

**Comment by termlen0** [comment]
> @termlen0 This is shaping up really nicely, I love it! One small thing missing: we need a sample vars file or at least entry for each new option we bring in, can you add this?
> 
> Also, while I would say @goetzrieger is a bit too cautious with the thoughts around control_nodes, we should at leas make sure that the other labs are working on it. So we need to fix these lines where stuff is install on control_nodes:
> 
>     * https://github.com/ansible/workshops/blob/2ff691f0471a0f3f2d9a4753dc395510b9e98471/provisioner/windows.yml#L17
> 
>     * https://github.com/ansible/workshops/blob/2ff691f0471a0f3f2d9a4753dc395510b9e98471/provisioner/roles/workshop_attendance/templates/workshop.sql.j2#L19
> 
>     * https://github.com/ansible/workshops/blob/2ff691f0471a0f3f2d9a4753dc395510b9e98471/provisioner/security.yml#L111
> 
> 
> Number one and three can just be rewritten to the ansible-1, but I am not sure about number 2, maybe @cloin can help?

I've made changes to address all 3 items. I've tested against the windows workshop. As for the sample vars, I'll update all existing samplevars files with the create_cluster boolean var and set it to "no" by default.

**Comment by liquidat** [comment]
recheck

**Comment by liquidat** [comment]
recheck

**Comment by liquidat** [comment]
We have builds failing.

**First, RHEL verify is failing:**

```
[2020-08-07T18:08:43.603Z] TASK [Test access by exporting assets] *****************************************
[2020-08-07T18:08:43.882Z] skipping: [student1-node1]
[2020-08-07T18:08:43.882Z] skipping: [student1-node2]
[2020-08-07T18:08:43.882Z] skipping: [student1-node3]
[2020-08-07T18:08:43.882Z] skipping: [student2-node1]
[2020-08-07T18:08:43.882Z] skipping: [student2-node2]
[2020-08-07T18:08:44.142Z] skipping: [student2-node3]
[2020-08-07T18:08:45.510Z] FAILED - RETRYING: Test access by exporting assets (60 retries left).
[...]
[2020-08-07T18:12:40.901Z] FAILED - RETRYING: Test access by exporting assets (1 retries left).
[2020-08-07T18:12:44.177Z] fatal: [student2-ansible-1]: FAILED! => changed=false 
[2020-08-07T18:12:44.177Z]   assets: null
[2020-08-07T18:12:44.177Z]   attempts: 60
[2020-08-07T18:12:44.177Z]   message: |-
[2020-08-07T18:12:44.177Z]     There was a network error of some kind trying to connect to Tower.
[2020-08-07T18:12:44.177Z]   
[2020-08-07T18:12:44.177Z]     The most common  reason for this is a settings issue; is your "host" value in `tower-cli config` correct?
[2020-08-07T18:12:44.177Z]     Right now it is: "student2-1.tqe-rhel-tower371-PR-891-32.rhdemo.io".
[2020-08-07T18:12:44.177Z]   msg: Receive Failed
[2020-08-07T18:12:44.433Z] fatal: [student1-ansible-1]: FAILED! => changed=false 
[2020-08-07T18:12:44.433Z]   assets: null
[2020-08-07T18:12:44.433Z]   attempts: 60
[2020-08-07T18:12:44.433Z]   message: |-
[2020-08-07T18:12:44.433Z]     There was a network error of some kind trying to connect to Tower.
[2020-08-07T18:12:44.433Z]   
[2020-08-07T18:12:44.433Z]     The most common  reason for this is a settings issue; is your "host" value in `tower-cli config` correct?
[2020-08-07T18:12:44.433Z]     Right now it is: "student1-1.tqe-rhel-tower371-PR-891-32.rhdemo.io".
[2020-08-07T18:12:44.433Z]   msg: Receive Failed
```
I think we need to modify this line in the testing script:
https://github.com/ansible/workshops/blob/916bc629d7e079dfd09f7f4eadf99ecb2c0154eb/provisioner/tests/rhel_verify.yml#L21

**Second, security deployment fails:**

```
[2020-08-07T17:38:26.714Z] TASK [manage_ec2_instances : provision workshop instances] *********************
[2020-08-07T17:38:26.714Z] ERROR! Unexpected Exception, this is probably a bug: expected str, bytes or os.PathLike object, not NoneType
```
This would be this line:
https://github.com/ansible/workshops/blob/916bc629d7e079dfd09f7f4eadf99ecb2c0154eb/provisioner/roles/manage_ec2_instances/tasks/provision.yml#L87
Honestly, I have no idea what is going on. I will try to provision on my own from this branch and see if I can replicate the problem.

**Comment by goetzrieger** [comment]
Easiest fix for RHEL verify fail IMO (if we want to stay with Ajay's naming convention):

file: provisioner/tests/rhel_verify.yml
```
tower_host: "{{ inventory_hostname|regex_replace('-ansible-1', '') }}.{{ workshop_name }}.rhdemo.io"
[...]
when: '"ansible-1" in inventory_hostname'
```

Tested with cluster and non-cluster RHEL WS.

**Comment by liquidat** [comment]
@termlen0 Can you please include @goetzrieger 's patch and also rebase? After the rebase I can track down the new bug, but right now without a rebase it is rather hard.

**Comment by termlen0** [comment]
Will do. Last 2 weeks were tough. Will try and get this in this week. Thanks!

**Comment by liquidat** [comment]
Build still fails here, this time with more information:
```
TASK [manage_ec2_instances : provision workshop instances] ***********************************************************************************************************************************
task path: /home/rwolters/gits/github/termlen0-linklight/provisioner/roles/manage_ec2_instances/tasks/provision.yml:47
ERROR! Unexpected Exception, this is probably a bug: expected str, bytes or os.PathLike object, not NoneType
the full traceback was:

Traceback (most recent call last):
  File "/home/rwolters/development/venv_ansible_2.9/bin/ansible-playbook", line 123, in <module>
    exit_code = cli.run()
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/cli/playbook.py", line 127, in run
    results = pbex.run()
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/executor/playbook_executor.py", line 169, in run
    result = self._tqm.run(play=play)
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/executor/task_queue_manager.py", line 241, in run
    play_return = strategy.run(iterator, play_context)
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/plugins/strategy/linear.py", line 359, in run
    new_blocks = self._load_included_file(included_file, iterator=iterator)
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/plugins/strategy/__init__.py", line 890, in _load_included_file
    block_list = load_list_of_blocks(
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/playbook/helpers.py", line 70, in load_list_of_blocks
    Block.load(
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/playbook/block.py", line 94, in load
    return b.load_data(data, variable_manager=variable_manager, loader=loader)
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/playbook/base.py", line 235, in load_data
    self._attributes[target_name] = method(name, ds[name])
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/playbook/block.py", line 122, in _load_block
    return load_list_of_tasks(
  File "/home/rwolters/development/venv_ansible_2.9/lib64/python3.8/site-packages/ansible/playbook/helpers.py", line 191, in load_list_of_tasks
    parent_include_dir = os.path.dirname(templar.template(parent_include.args.get('_raw_params')))
  File "/usr/lib64/python3.8/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

**Comment by liquidat** [comment]
So, this looks like an evil Ansible bug. For me, this change made it working:
```
diff --git a/provisioner/provision_lab.yml b/provisioner/provision_lab.yml
index 2f86777f..50487323 100644
--- a/provisioner/provision_lab.yml
+++ b/provisioner/provision_lab.yml
@@ -18,15 +18,14 @@
   connection: local
   become: false
   gather_facts: false
-  tasks:
+  pre_tasks:
     - name: Cluster nodes
       set_fact:
         control_nodes: 4
       when: create_cluster is defined and create_cluster|bool
 
-    - name: Manage EC2
-      include_role:
-        name: manage_ec2_instances
+  roles:
+    - manage_ec2_instances
 
 - name: wait for all nodes to have SSH reachability
   hosts: "managed_nodes:control_nodes:attendance"
```
@termlen0 You mentioned in chat you don't see this behavior. I tested with ansible version 2.9.9 and 2.9.12, both times I see the same problem.
Any reason why we could not adopt the change mentioned above? Besides the fact that it is ugly?

**Comment by liquidat** [comment]
@termlen0 Security deployment is mostly fine. The RHLE verify script still fails as mentioned above, @goetzrieger [had a patch](https://github.com/ansible/workshops/pull/891#issuecomment-672083595), did you include it?

There is an error with the security verify as well, but I'd like a recheck to be sure that this is not a fluke.

**Comment by termlen0** [comment]
> @termlen0 Security deployment is mostly fine. The RHLE verify script still fails as mentioned above, @goetzrieger [had a patch](https://github.com/ansible/workshops/pull/891#issuecomment-672083595), did you include it?
> 
> There is an error with the security verify as well, but I'd like a recheck to be sure that this is not a fluke.

Just committed.

**Comment by liquidat** [comment]
@termlen0 We missed something in the security workshop: the checkpoint stuff isn't even called and thus the test fails. Can you please add this patch?
```
diff --git a/provisioner/roles/cp_setup/tasks/main.yml b/provisioner/roles/cp_setup/tasks/main.yml
index 60b2105f..1a68a42a 100644
--- a/provisioner/roles/cp_setup/tasks/main.yml
+++ b/provisioner/roles/cp_setup/tasks/main.yml
@@ -1,7 +1,7 @@
 ---
 - name: login, get SID
   uri:
-    url: "https://{{ hostvars[inventory_hostname|regex_replace('ansible', 'checkpoint_mgmt')]['private_ip'] }}/web_api/login"
+    url: "https://{{ hostvars[inventory_hostname|regex_replace('ansible-1', 'checkpoint_mgmt')]['private_ip'] }}/web_api/login"
     method: POST
     body:
       user: admin
@@ -15,7 +15,7 @@
 
 - name: Add NGFW to MGMT
   uri:
-    url: "https://{{ hostvars[inventory_hostname|regex_replace('ansible', 'checkpoint_mgmt')]['private_ip'] }}/web_api/add-simple-gateway"
+    url: "https://{{ hostvars[inventory_hostname|regex_replace('ansible-1', 'checkpoint_mgmt')]['private_ip'] }}/web_api/add-simple-gateway"
     validate_certs: false
     method: POST
     headers:
@@ -24,7 +24,7 @@
     body_format: json
     body:
       name: myngfw
-      ip-address: "{{ hostvars[inventory_hostname|regex_replace('ansible', 'checkpoint_gw')]['private_ip'] }}"
+      ip-address: "{{ hostvars[inventory_hostname|regex_replace('ansible-1', 'checkpoint_gw')]['private_ip'] }}"
       one-time-password: admin123
       firewall: true
       version: R80.30
@@ -34,7 +34,7 @@
           anti-spoofing-settings:
             action: prevent
           name: "eth0"
-          ip-address: "{{ hostvars[inventory_hostname|regex_replace('ansible', 'checkpoint_gw')]['private_ip'] }}"
+          ip-address: "{{ hostvars[inventory_hostname|regex_replace('ansible-1', 'checkpoint_gw')]['private_ip'] }}"
           network-mask: "255.255.0.0"
           ipv4-mask-length: 16
           security-zone: false
@@ -45,7 +45,7 @@
           anti-spoofing-settings:
             action: prevent
           name: "eth1"
-          ip-address: "{{ hostvars[inventory_hostname|regex_replace('ansible', 'checkpoint_gw')]['private_ip2'] }}"
+          ip-address: "{{ hostvars[inventory_hostname|regex_replace('ansible-1', 'checkpoint_gw')]['private_ip2'] }}"
           network-mask: "255.255.0.0"
           ipv4-mask-length: 16
           security-zone: false
@@ -53,7 +53,7 @@
 
 - name: Publish
   uri:
-    url: "https://{{ hostvars[inventory_hostname|regex_replace('ansible', 'checkpoint_mgmt')]['private_ip'] }}/web_api/publish"
+    url: "https://{{ hostvars[inventory_hostname|regex_replace('ansible-1', 'checkpoint_mgmt')]['private_ip'] }}/web_api/publish"
     validate_certs: false
     method: POST
     headers:
@@ -67,7 +67,7 @@
   ec2_instance_info:
     region: "{{ ec2_region }}"
     filters:
-      "tag:Name": "{{ inventory_hostname|regex_replace('ansible', 'checkpoint_gw') }}"
+      "tag:Name": "{{ inventory_hostname|regex_replace('ansible-1', 'checkpoint_gw') }}"
       "instance-state-name": running
   register: gw_inst
   delegate_to: localhost
diff --git a/provisioner/security.yml b/provisioner/security.yml
index 53dcbc38..c3e4b25b 100644
--- a/provisioner/security.yml
+++ b/provisioner/security.yml
@@ -108,6 +108,6 @@
     - role: cp_fix_mgmt
 
 - name: SETUP CHECKPOINT ENVIRONMENT
-  hosts: ansible-1
+  hosts: '*ansible-1'
   roles:
     - role: cp_setup
```

With this patch my tests are all good.

**Pull request merged**